dive_core/perf_metrics_data.cpp

```diff
@@ -0,0 +1,248 @@
+    errno = 0;
+    float val = std::strtof(start, &end);
+    if (errno == ERANGE || *end != '\0' || start == end)
+    {
+        return false;
+    }
+    out = val;
+    return true;
+}
+
+bool GetTrimmedLine(std::ifstream& file, std::string& line)
+{
+    if (!std::getline(file, line))
+        return false;
+    Trim(line);
+    return true;
+}
+
+bool GetTrimmedField(std::stringstream& ss, std::string& field, char delimiter = ',')
+{
+    if (!std::getline(ss, field, delimiter))
+        return false;
+    Trim(field);
+    return true;
+}
+
+}  // namespace
+
+std::unique_ptr<PerfMetricsData> PerfMetricsData::LoadFromCsv(
+const std::filesystem::path& file_path,
+const AvailableMetrics&      available_metrics)
+{
+    std::vector<std::string>       metric_names;
+    std::vector<const MetricInfo*> metric_infos;
+    std::vector<PerfMetricsRecord> records;
+
+    std::ifstream file(file_path);
+    if (!file.is_open())
+    {
+        std::cerr << "Failed to open file: " << file_path << std::endl;
+        return nullptr;
+    }
+
+    std::string line;
+    // Read header line
+    if (!GetTrimmedLine(file, line) || line.empty())
+    {
+        return nullptr;
+    }
+
+    std::stringstream header_ss(line);
+    std::string       header_field;
+    int               column_index = 0;
+    while (GetTrimmedField(header_ss, header_field, ','))
+    {
+        if (column_index < kFixedHeaders.size())
+        {
+            if (header_field != kFixedHeaders[column_index])
+                return nullptr;
+        }
+        else
+        {
+            metric_names.push_back(header_field);
+            metric_infos.push_back(available_metrics.GetMetricInfo(header_field));
+        }
+        column_index++;
+    }
+    if (column_index < kFixedHeaders.size())
+        return nullptr;
+
+    while (GetTrimmedLine(file, line))
+    {
+        std::stringstream        ss(line);
+        std::string              field;
+        std::vector<std::string> fields;
+        while (GetTrimmedField(ss, field, ','))
+        {
+            fields.push_back(field);
+        }
+
+        if (fields.size() != kFixedHeaders.size() + metric_names.size())
+        {
+            continue;  // Skip malformed lines
+        }
+
+        PerfMetricsRecord record{};
+        uint32_t          draw_type = 0, lrz_state = 0;
+        if (!SafeStoull(fields[0], record.m_context_id) ||
+            !SafeStoull(fields[1], record.m_process_id) ||
+            !SafeStoull(fields[2], record.m_frame_id) ||
+            !SafeStoull(fields[3], record.m_cmd_buffer_id) ||
+            !SafeStoul(fields[4], record.m_draw_id) || !SafeStoul(fields[5], draw_type) ||
+            !SafeStoul(fields[6], record.m_draw_label) ||
+            !SafeStoull(fields[7], record.m_program_id) || !SafeStoul(fields[8], lrz_state))
+        {
+            continue;  // Skip lines with parsing errors
+        }
+        record.m_draw_type = static_cast<uint8_t>(draw_type);
```

Comment:
potential overflow by converting from uint32_t to uint8_t. (which could be caught by SafeConvertFromString() with the std::numeric_limits<T> check)